### PR TITLE
Special case the @CONST, @ABSTRACT, and @IMPLEMENTS annotations.

### LINKED ISSUE #98: Handle @CONST, @IMPLEMENTS, @ABSTRACT

### PR DESCRIPTION
**Pull request by mprobst**
Fixes #98.

### CODE REVIEW

**Comment by mprobst** [comment]
Sorry about the size of it. And overall it's pretty ugly hacks - but once we're out of AtScript, we can drop some of it again.

File: main.ts [hunk 1]
```diff
@@ -97,11 +102,42 @@ class Translator {
...
+        this.emit('implements');
+      }
+      this.emit(implIfs.join(' , '));
+    }
     this.emit('{');
     this.visitEachIfPresent(decl.members);
     this.emit('}');
   }
 
+  /** Returns parameters to @IMPLEMENTS, returns their identifiers. */
```
Comment (alexeagle) [inline main.ts]: this doesn't read right

File: main.ts [hunk 2]
```diff
@@ -97,11 +102,42 @@ class Translator {
...
+  getImplementsDecorators(decorators: ts.NodeArray<ts.Decorator>): string[] {
+    var interfaces = [];
+    if (!decorators) return interfaces;
```
Comment (alexeagle) [inline main.ts]: interesting that you don't have to declare the type of interfaces

File: main.ts [hunk 3]
```diff
@@ -111,6 +147,58 @@ class Translator {
...
+        return;
+      }
+      if (name === 'IMPLEMENTS') {
+        // Ignore @IMPLEMENTS - it's handled above in visitClassLike.
+        return;
+      }
+      this.emit('@');
+      this.visit(d.expression);
+    });
+    if (isAbstract) this.emit('abstract');
```
Comment (alexeagle) [inline main.ts]: could emit inside the loop, unless we are really concerned that CONST or ABSTRACT appears more than once

File: main.ts [hunk 4]
```diff
@@ -174,21 +262,92 @@ class Translator {
     this.visit(expr);
   }
 
-  /** Searches for super() calls and emits an initializer for it if found. */
-  maybeEmitInitializerListForSuperCall(body: ts.Block) {
-    if (!body || body.statements.length === 0) return;
+  /**
+   * Handles constructor initializer lists and bodies.
+   *
+   * <p>Dart's super() ctor calls have to be moved to the constructors initializer list, and `const`
```
Comment (alexeagle) [inline main.ts]: note
https://github.com/angular/angular/issues/70

(I don't think you're fixing here?)

File: test/DartTest.ts [hunk 5]
```diff
@@ -537,5 +565,5 @@ export function translateSource(contents: string): string {
     var first = program.getSyntacticDiagnostics()[0];
     throw new Error(`${first.start}: ${first.messageText} in ${contents}`);
   }
-  return main.translateProgram(program);
+  return main.translateProgram(program, {failFast: failFast});
```
Comment (alexeagle) [inline test/DartTest.ts]: is this the place where you can do the shorthand {failFast} ?

**Comment by alexeagle** [comment]
Nice, if ugly. LGTM

File: main.ts (hunk 1 above)
Comment (mprobst) [inline main.ts]: done.

File: main.ts (hunk 2 above)
Comment (mprobst) [inline main.ts]: That's odd indeed. I checked the spec:
https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3.10.3

This sounds like it's a bug, at least I cannot find language in the spec to the point of converting `Array<any>` to `Array<string>`.

File: main.ts (hunk 3 above)
Comment (mprobst) [inline main.ts]: Done. I don't think TS allows duplicated annotations.

File: main.ts (hunk 4 above)
Comment (mprobst) [inline main.ts]: Nope, this is just handling the initializer stuff, ctor inheritance is a different can of worms.

File: test/DartTest.ts (hunk 5 above)
Comment (mprobst) [inline test/DartTest.ts]: Done. I'm still not sure I like that syntax.

**Comment by mprobst** [comment]
Yeah, these AST transformations always escalate quickly. But stuff got done, which is always nice.

**Pull request merged**